Write delete predicate into RowsetMeta upon upgrade from Doris-0.10 to Doris-0.11

chaoyli opened the pull request:
If delete predicate exists in meta in Doris-0.10, all of this predicates should
be remained. There is an confused place in Doris-0.10. The delete predicate
only exists in OLAPHeaderMessage and PPendingDelta, not in PDelta.
This trick results this bug.

Commit:
Write delete predicate into RowsetMeta upon upgrade from Doris-0.10 to Doris-0.11

If delete predicate exists in meta in Doris-0.10, all of this predicates should
be remained. There is an confused place in Doris-0.10. The delete predicate
only exists in OLAPHeaderMessage and PPendingDelta, not in PDelta.
This trick results this bug.

chaoyli commented:
#3045

imay commented on be/src/olap/olap_snapshot_converter.cpp:
```diff
@@ -89,27 +93,27 @@ OLAPStatus OlapSnapshotConverter::to_olap_header(const TabletMetaPB& tablet_meta
 }
 
-OLAPStatus OlapSnapshotConverter::to_tablet_meta_pb(const OLAPHeaderMessage& olap_header,
+OLAPStatus OlapSnapshotConverter::to_tablet_meta_pb(OLAPHeaderMessage* olap_header,
```
This function name indicates filling content from olap_header. It is not a good idea to change olap_header's content.
And comment about what will be done in this function is needed.

chaoyli replied:
Because PDelta can not corresponds with RowsetMeta. 
So before calling convert_to_rowset_meta, I add delete predicate to PDelta to store delete predicate.

imay commented on be/src/olap/olap_snapshot_converter.h:
```diff
@@ -47,7 +47,7 @@ class OlapSnapshotConverter {
     // pending delta is not in tablet meta any more, so that convert pending delta to rowset and add it to pending rowsets
     // as a return value
-    OLAPStatus to_tablet_meta_pb(const OLAPHeaderMessage& olap_header, TabletMetaPB* tablet_meta_pb, 
+    OLAPStatus to_tablet_meta_pb(OLAPHeaderMessage* olap_header, TabletMetaPB* tablet_meta_pb,
```
Originally, this function only converts the header to meta.
Can you implement another function to finish new added operation to make it less complicate?

imay commented on be/src/olap/olap_snapshot_converter.cpp:
```diff
@@ -131,11 +136,23 @@ OLAPStatus OlapSnapshotConverter::to_tablet_meta_pb(const OLAPHeaderMessage& ola
+                && temp_delta.start_version() == del_pred.version()) {
+                    DeletePredicatePB* delete_condition = temp_delta.mutable_delete_condition();
+                    *delete_condition = del_pred;
+                }
```
align

imay replied:
```suggestion
        if (temp_delta.start_version() == temp_delta.end_version()) {
            for (auto& del_pred : delete_conditions) {
                if (temp_delta.start_version() == del_pred.version()) {
                    DeletePredicatePB* delete_condition = temp_delta.mutable_delete_condition();
                    *delete_condition = del_pred;
                }
```

imay commented on be/src/olap/olap_snapshot_converter.cpp:
```diff
@@ -150,7 +167,15 @@ OLAPStatus OlapSnapshotConverter::to_tablet_meta_pb(const OLAPHeaderMessage& ola
+                && temp_inc_delta.start_version() == del_pred.version()) {
+                    DeletePredicatePB* delete_condition = temp_inc_delta.mutable_delete_condition();
+                    *delete_condition = del_pred;
+                }
```
align

Commit:
Write delete predicate into RowsetMeta upon upgrade from Doris-0.10 to Doris-0.11

If delete predicate exists in meta in Doris-0.10, all of this predicates should
be remained. There is an confused place in Doris-0.10. The delete predicate
only exists in OLAPHeaderMessage and PPendingDelta, not in PDelta.
This trick results this bug.

imay reviewed (approved):
LGTM
